erts/emulator/beam/erl_bif_lists.c

```diff
@@ -976,7 +990,31 @@ static int subtract_start(Process *p, Eterm lhs, Eterm rhs,
 static Eterm subtract(Export *bif_entry, BIF_ALIST_2) {
     Eterm lhs = BIF_ARG_1, rhs = BIF_ARG_2;
 
-    if ((is_list(lhs) || is_nil(lhs)) && (is_list(rhs) || is_nil(rhs))) {
+    if (is_nil(lhs) && is_list(rhs)) {
+        /* Fast-lane when the `lhs` is nil: return nil.
+         * First check that `rhs` is not an improper list. */
+        Eterm list = rhs;
+        while (is_list(list)) {
+            list = CDR(list_val(list));
+        }
+        if (is_not_nil(list)) {
+            BIF_ERROR(BIF_P, BADARG);
+        }
+
+        return lhs;
+    } else if (is_list(lhs) && is_nil(rhs)) {
+        /* Similar fast-lane when the rhs is nil: return the lhs list.
+         * First check that `lhs` is not an improper list. */
+        Eterm list = lhs;
+        while (is_list(list)) {
+            list = CDR(list_val(list));
+        }
+        if (is_not_nil(list)) {
+            BIF_ERROR(BIF_P, BADARG);
+        }
+
+        return lhs;
+    } else if ((is_list(lhs) || is_nil(lhs)) && (is_list(rhs) || is_nil(rhs))) {
```

Comment:
This also needs to yield.

It's probably easier to do this at the end of `SUBTRACT_STAGE_LEN_RHS`:  if `context->rhs_remaining` is 0, set `context->result` to LHS and return.